pkg/ipamd/ipamd.go

```diff
@@ -579,66 +609,70 @@ func (c *IPAMContext) tryUnassignIPsFromAll() {
 			}
 
 			// Delete IPs from datastore
-			var deletedIPs []string
+			var deletedIPsOrPrefixes []string
 			for _, toDelete := range ips {
 				// Don't force the delete, since a freeable IP might have been assigned to a pod
 				// before we get around to deleting it.
-				err := c.dataStore.DelIPv4AddressFromStore(eniID, toDelete, false /* force */)
+				err := c.dataStore.CleanupDataStore(eniID, toDelete, false /* force */, c.enableIpv4PrefixDelegation)
+
 				if err != nil {
 					log.Warnf("Failed to delete IP %s on ENI %s from datastore: %s", toDelete, eniID, err)
 					ipamdErrInc("decreaseIPPool")
 					continue
 				} else {
-					deletedIPs = append(deletedIPs, toDelete)
+					deletedIPsOrPrefixes = append(deletedIPsOrPrefixes, toDelete)
 				}
-			}
 
-			// Deallocate IPs from the instance if they aren't used by pods.
-			if err := c.awsClient.DeallocIPAddresses(eniID, deletedIPs); err != nil {
-				log.Warnf("Failed to decrease IP pool by removing IPs %v from ENI %s: %s", deletedIPs, eniID, err)
-			} else {
-				log.Debugf("Successfully decreased IP pool by removing IPs %v from ENI %s", deletedIPs, eniID)
+				// Deallocate IPs from the instance if they aren't used by pods.
+				if err = c.awsClient.DeallocIPAddresses(eniID, deletedIPsOrPrefixes, false); err != nil {
+					log.Warnf("Failed to decrease pool by removing IPs %v from ENI %s: %s", deletedIPsOrPrefixes, eniID, err)
+				} else {
+					log.Debugf("Successfully decreased pool by removing IPs %v from ENI %s", deletedIPsOrPrefixes, eniID)
+				}
 			}
 
 			// Track the last time we unassigned IPs from an ENI. We won't reconcile any IPs in this cache
 			// for at least ipReconcileCooldown
-			c.reconcileCooldownCache.Add(deletedIPs)
+			c.reconcileCooldownCache.Add(deletedIPsOrPrefixes)
 		}
 	}
 }
 
-// findFreeableIPs finds and returns IPs that are not assigned to Pods but are attached
+// findFreeableIPsOrPrefixes finds and returns IPs that are not assigned to Pods but are attached
 // to ENIs on the node.
-func (c *IPAMContext) findFreeableIPs(eni string) ([]string, error) {
-	freeableIPs := c.dataStore.FreeableIPs(eni)
-
+func (c *IPAMContext) findFreeableIPsOrPrefixes(eni string) ([]string, error) {
```

Comment:
Same feedback as `tryUnassignIPsOrPrefixesFromAll `

Reply:
Same as above, discussed about this. Will take it up in the following PR.